npm/private/npm_translate_lock_state.bzl

```diff
@@ -281,6 +293,31 @@ WARNING: Implicitly using package.json file `{package_json}` since the `{pnpm_lo
                 fail(msg)
             _copy_input_file(priv, rctx, label_store, package_json_key)
 
+    # pnpm.patchedDependencies patch files
+    root_package_json = _read_root_package_json(rctx, label_store)
```

Comment:
it looks like you're calling _read_root_package_json in two spots. You could cache the read to eliminate the duplicate file read & json parsing

Reply:
Added caching inside `_read_root_package_json`.